src/plugins/flow.js

```diff
@@ -1162,6 +1162,7 @@ export default function (instance) {
         let state = this.state.clone();
         try {
           let returnType = this.flowParseTypeAnnotation();
+          if (this.canInsertSemicolon()) this.unexpected();
```

Comment:
Argh, true.

This try catch is there because when we encounter a colon at this spot we are not yet sure about a.) if it is an arrow at all b.) if the colon might be from a ternary operator (something like this i think `a ? b: c => {}`). So if a SyntaxError is thrown while parsing the return type we reset the state to what it was before we started parsing the type and let the regular parsing function of babylon handle it.